src/VisualStudio/Core/Def/Implementation/Preview/PreviewUpdater.cs

```diff
@@ -52,34 +50,65 @@ public void UpdateView(Document document, SpanChange spanSource)
             Tagger.OnTextBufferChanged();
         }
 
-        private void UpdateBuffer(Document document, SpanChange spanSource, out SourceTextContainer container, out Document documentBackedByTextBuffer)
+        private void UpdateBuffer(TextDocument document, SpanChange spanSource, bool isAdditionalDocument, out SourceTextContainer container, out TextDocument documentBackedByTextBuffer)
         {
             if (_previewWorkspace != null)
             {
-                _previewWorkspace.CloseDocument(_currentDocument, _previewWorkspace.CurrentSolution.GetDocument(_currentDocument).GetTextAsync().Result);
-
-                // Put the new document into the current preview solution
-                var updatedSolution = _previewWorkspace.CurrentSolution.WithDocumentText(document.Id, document.GetTextAsync().Result);
-                var updatedDocument = updatedSolution.GetDocument(document.Id);
-
-                ApplyDocumentToBuffer(updatedDocument, spanSource, out container, out documentBackedByTextBuffer);
+                // Close the current document in preview solution.
+                if (_isCurrentAdditionalDocument)
```

Comment:
both flags _isCurrentAdditionalDocument and isAdditionalDocument look like can be figured out by _currentDocument and document. 

since retrieving that information from the documents is not expensive. having less flag probably better I think.

also, it would be nice if there is a extension method that 
GetDocumentKind(TextDocument document)
{
if (document is Document)
{
return "SomeKindOfEnum?".Document;
}

// not sure whether there is a better way?
if (document.Solution.GetAdditionalDocument(document.Id) == document)
{
return "Additional..."
}

return "Text..."
}